src/main/java/com/google/devtools/build/lib/vfs/bazel/Blake3MessageDigest.java

```diff
@@ -0,0 +1,229 @@
+package com.google.devtools.build.lib.vfs.bazel;
+
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.common.hash.Funnel;
+import com.google.common.hash.HashCode;
+import com.google.common.hash.Hasher;
+import com.google.errorprone.annotations.CanIgnoreReturnValue;
+import java.nio.ByteBuffer;
+import java.nio.charset.Charset;
+import java.security.DigestException;
+import java.security.MessageDigest;
+
+public final class Blake3MessageDigest extends MessageDigest implements Hasher {
+  // These constants match the native definitions in:
+  // https://github.com/BLAKE3-team/BLAKE3/blob/master/c/blake3.h
+  public static final int KEY_LEN = 32;
+  public static final int OUT_LEN = 32;
+
+  // To reduce the number of calls made via JNI, buffer up to this many bytes.
+  // If a call to "hash()" is made and less than this much data has been
+  // written, a single JNI call will be made that initializes, hashes, and
+  // cleans up the hasher, rather than making separate calls for each operation.
+  public static final int ONESHOT_THRESHOLD = 8 * 1024;
+  private ByteBuffer buffer = ByteBuffer.allocate(ONESHOT_THRESHOLD);
+
+  private long hasher = -1;
+  private boolean isDone;
+
+  public Blake3MessageDigest() {
+    super("BLAKE3");
+    isDone = false;
+  }
+
+  private void flush() {
+    if (hasher == -1) {
+      hasher = Blake3JNI.allocate_and_initialize_hasher();
+    }
+
+    if (buffer.position() > 0) {
+      Blake3JNI.blake3_hasher_update(hasher, buffer.array(), buffer.position());
+      buffer.clear();
+    }
+  }
+
+  public void engineUpdate(byte[] data, int offset, int length) {
+    while (length > 0) {
```

Comment:
One optimization here is we could avoid copying if the size of `data` is larger than `ONESHOT_THRESHOLD`. But this is totally optional and we could leave that for a separate PR.

Reply:
Yeah I was trying to keep this simple for now -- enough complexity already :) Agree on deferring this for a followup (where we measure perf).